keylime/secure_mount.py

```diff
@@ -85,6 +86,13 @@ def mount():
 
 def umount():
     """Umount all the devices mounted by Keylime."""
+
+    # Make sure we leave tmpfs dir empty even if we did not mount it
+    try:
+        shutil.rmtree(mount())
+    except OSError:
+        pass
```

Comment:
I am not sure that this is OK. For one, if the user decided not to use `tmpfs` maybe really do not want to remove the payload (to debug, inspect or audit)

Also, we are calling `mount`, with is totally confusing because is a function that have a side effect (I know that in the happy path is idempotent, but is scary).

And finally we should at least `log.error()`.

But IMHO all this change should not be present?

Reply:
I was wondering why we bother with mounting tmpfs at all when one can just remove a whole directory upon shutdown. Because if the shutdown doesn't happen (kill -9) then we are in the same situation as not removing the directory and files still lay around. So I am still wondering about the tmpfs.

For simplicity reasons I chose `shutil.rmtree()` to remove everything under and including the `secure` directory. We will get an error when this function tries to remove the mount point `secure` since tmpfs is still mounted. I don't think we should log this. Now using this (best-effort):

```
    # Make sure we leave tmpfs dir empty even if we did not mount it or
    # if we cannot unmount it. Ignore errors while deleting. The deletion
    # of the 'secure' directory will result in an error since it's a mount point.
    shutil.rmtree(mount(), ignore_errors=True)
```

Reply:
I see, but I still will like to drop this change and postpone it to a different PR, and focus this one only in the downgrade case.

I think that the `tmpfs` analysis can be more complicated.

Reply:
But this is part of downgrading since we cannot unmount anymore but have to delete the contents of this directory to achieve a similar effect. I think it would be a bug in this PR not to do this.

Reply:
> I think that the `tmpfs` analysis can be more complicated.

Not sure. It's mounted at the beginning, filled with files, unmounted at the end and files are gone.

Reply:
if `config.SECURE_MOUNT` is true then we should remove the files otherwise we can keep them for debugging purposes.

Reply:
> But this is part of downgrading since we cannot unmount anymore but have to delete the contents of this directory to achieve a similar effect. I think it would be a bug in this PR not to do this.

Exactly. Downgrading privileges makes umount impossible if secure is not part of fstab (with the user parameter). That is exactly we should make the mount / umount outside the agent.

If we insist of doing the mount / umount in the agent, we should the do mount after the drop of privileges (like was before). If the mount can be done so can be umounted.

Reply:
> If we insist of doing the mount / umount in the agent, we should the do mount after the drop of privileges (like was before). If the mount can be done so can be umounted.

That's not possible:
```
$ whoami
stefanb
$ mkdir foo
$ mount -t tmpfs ./foo ./foo/
mount: /home/stefanb/dev/keylime/foo: must be superuser to use mount.
```

Reply:
Btw, Linux kernel should be able to do a lazy unmount.

Reply:
> > If we insist of doing the mount / umount in the agent, we should the do mount after the drop of privileges (like was before). If the mount can be done so can be umounted.
> 
> That's not possible:

It is possible. From `man 8 mount`:

```
   Non-superuser mounts
       Normally, only the superuser can mount filesystems. However, when fstab contains the user option on
       a line, anybody can mount the corresponding filesystem.

       Thus, given a line

          /dev/cdrom /cd iso9660 ro,user,noauto,unhide

       any user can mount the iso9660 filesystem found on an inserted CDROM using the command:

          mount /cd
```

Reply:
> Btw, Linux kernel should be able to do a lazy unmount.

Reading more about it now, most likely it won't be the safest approach.